### PR TITLE
[SPARK-24266][K8S][2.4] Restart the watcher when we receive a version changed from k8s

### PR DESCRIPTION
**Pull request by jkleckner**
### What changes were proposed in this pull request?

This patch processes the HTTP Gone event and restarts the pod watcher.

### Why are the changes needed?

This is a backport of PR #28423 to branch-2.4.
The reasons are explained in SPARK-24266 that spark jobs using the k8s resource scheduler may hang.

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

Manually.

### CODE REVIEW

**Comment by jkleckner** [comment]
@shockdm Please check this over.

@dongjoon-hyun Maybe we can get this into 2.4 ?

My force-push prevented the old PR from being reopened it seems.

**Comment by dongjoon-hyun** [comment]
ok to test

**Comment by jkleckner** [comment]
An image of just the spark build was created with this job [1].

We are using this image starting today in our dev cluster.

[1] https://gitlab.com/jkleckner/spark/-/pipelines/212962140

**Comment by jkleckner** [comment]
This version of Spark has been running find since the last 3 days in our cluster.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala [hunk 1]
```diff
@@ -28,8 +29,10 @@ import org.apache.spark.SparkException
 import org.apache.spark.internal.Logging
 import org.apache.spark.util.ThreadUtils
 
+
 private[k8s] trait LoggingPodStatusWatcher extends Watcher[Pod] {
-  def awaitCompletion(): Unit
+  def reset(): Unit
+  def watchOrStop(sId: String): Boolean
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala:35]: Could you preserve the order from the [original patches](https://github.com/apache/spark/pull/28423/files#diff-ff85a8d70a121a39181ed5ddb1005f87bf677acbe0849defc9fc843bcef62df2R31) (master/branch-3.0)?
```scala
def watchOrStop(submissionId: String): Boolean
def reset(): Unit
```

The function declaration order is different. The parameter name is not the same here (sId != submissionId)

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala [hunk 2]
```diff
@@ -42,9 +45,12 @@ private[k8s] trait LoggingPodStatusWatcher extends Watcher[Pod] {
  */
 private[k8s] class LoggingPodStatusWatcherImpl(
     appId: String,
-    maybeLoggingInterval: Option[Long])
+    maybeLoggingInterval: Option[Long],
+    waitForCompletion: Boolean)
   extends LoggingPodStatusWatcher with Logging {
 
+  private var resourceTooOldReceived: Boolean = false
+  private var podCompleted = false
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala:53]: ditto. If possible, shall we follow the order and spacing in `master branch`.
- https://github.com/apache/spark/blob/master/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala#L45-L48

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala [hunk 3]
```diff
@@ -77,9 +83,19 @@ private[k8s] class LoggingPodStatusWatcherImpl(
...
   }
 
+  override def reset(): Unit = {
+    resourceTooOldReceived = false
+  }
+
   override def onClose(e: KubernetesClientException): Unit = {
-    logDebug(s"Stopping watching application $appId with last-observed phase $phase")
-    closeWatch()
+    logInfo(s"Stopping watching application $appId with last-observed phase $phase")
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala:91]: This is a regression because this is previous `logDebug`. (Of course, this is different from `master` branch, too).

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala [hunk 4]
```diff
@@ -77,9 +83,19 @@ private[k8s] class LoggingPodStatusWatcherImpl(
...
+    if (e != null && e.getCode==HTTP_GONE) {
+      resourceTooOldReceived = true
+      logInfo(s"Got HTTP Gone code, resource version changed in k8s api: $e")
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala:94]: ditto.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala [hunk 5]
```diff
@@ -77,9 +83,19 @@ private[k8s] class LoggingPodStatusWatcherImpl(
...
+    } else {
+      logInfo(s"Got proper termination code, closing watcher.")
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala:96]: Please remove this.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala [hunk 6]
```diff
@@ -134,13 +151,6 @@ private[k8s] class LoggingPodStatusWatcherImpl(
     }.mkString("")
   }
 
-  override def awaitCompletion(): Unit = {
-    podCompletedFuture.await()
-    logInfo(pod.map { p =>
-      s"Container final statuses:\n\n${containersDescription(p)}"
-    }.getOrElse("No containers were found in the driver pod."))
-  }
-
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala:143]: This removal looks like a part of independency PR instead of the part of SPARK-24266. Could you tell us why this is required and where this came from?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala [hunk 7]
```diff
@@ -177,4 +187,35 @@ private[k8s] class LoggingPodStatusWatcherImpl(
   private def formatTime(time: String): String = {
     if (time != null ||  time != "") time else "N/A"
   }
+
+  override def watchOrStop(sId: String): Boolean = if (waitForCompletion) {
+    logInfo(s"Patched Sept 8th: Waiting for application" +
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala:192]: `Patched Sept 8th:`?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala [hunk 8]
```diff
@@ -177,4 +187,35 @@ private[k8s] class LoggingPodStatusWatcherImpl(
...
+      s" ${appId} with submission ID $sId to finish...")
+    val interval = maybeLoggingInterval
+
+    synchronized {
+      while (!podCompleted && !resourceTooOldReceived) {
+        wait(interval.get)
+        logDebug(s"Application status for $appId (phase: $phase)")
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala:199]: This should be `logInfo`.

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/submit/ClientSuite.scala [hunk 9]
```diff
@@ -151,6 +151,7 @@ class ClientSuite extends SparkFunSuite with BeforeAndAfter {
     createdResourcesArgumentCaptor = ArgumentCaptor.forClass(classOf[HasMetadata])
     when(podOperations.create(FULL_EXPECTED_POD)).thenReturn(POD_WITH_OWNER_REFERENCE)
     when(namedPods.watch(loggingPodStatusWatcher)).thenReturn(mock[Watch])
+    when(loggingPodStatusWatcher.watchOrStop("default" + ":" + POD_NAME)).thenReturn(true)
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/submit/ClientSuite.scala:154]: Maybe, `kubernetesConf.namespace()` instead of `"default"`?

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
Hi, @jkleckner . 
This will be an official Apache Spark patch. Could you clean up something like `logInfo(s"Patched Sept 8th:` and make it consistent across Apache Spark branches?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala (hunk 6 above)
Comment (jkleckner) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala:143]: @shockdm Could you chime in on this one?

Comment (jkleckner) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala:143]: @dongjoon-hyun It does look like it originates from SPARK-28947 02c5b4f76337cc3901b8741887292bb4478931f3 which eliminated the future and was a rename.

Since this is a private trait, the logic should be completely self-contained and safe to remove.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala (hunk 1 above)
Comment (jkleckner) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala]: Done.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala (hunk 2 above)
Comment (jkleckner) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala:52]: Done.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala (hunk 3 above)
Comment (jkleckner) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala]: Done

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala (hunk 4 above)
Comment (jkleckner) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala]: Done.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala (hunk 5 above)
Comment (jkleckner) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala]: Done.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala (hunk 8 above)
Comment (jkleckner) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala:201]: Done.

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/submit/ClientSuite.scala (hunk 9 above)
Comment (jkleckner) [reply resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/submit/ClientSuite.scala]: Done.

**Comment by jkleckner** [comment]
@dongjoon-hyun Sorry for doing force-push.  I got used to gitlab which keeps track of history better.

**Comment by dongjoon-hyun** [comment]
Retest this please

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala (hunk 6 above)
Comment (shockdm) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala:143]: @jkleckner thank you for following up, that is correct. Sorry for the late response :(

**Comment by dongjoon-hyun** [comment]
Retest this please.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesClientApplication.scala [hunk 10]
```diff
@@ -133,29 +136,37 @@ private[spark] class Client(
...
+        kubernetesClient.pods().delete(createdDriverPod)
+        throw e
+    }
+    val sId = Seq(kubernetesConf.namespace(), driverPodName).mkString(":")
+    breakable {
+      while (true) {
+        val podWithName = kubernetesClient
+          .pods()
+          .withName(driverPodName)
+
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesClientApplication.scala:158]: Could you add the following comments like the other branches?
```
// Reset resource to old before we start the watch, this is important for race conditions
```

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesClientApplication.scala [hunk 11]
```diff
@@ -133,29 +136,37 @@ private[spark] class Client(
...
+        watcher.reset()
+
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesClientApplication.scala:160]: Let's remove this empty line.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesClientApplication.scala [hunk 12]
```diff
@@ -133,29 +136,37 @@ private[spark] class Client(
...
+        watch = podWithName.watch(watcher)
+
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesClientApplication.scala:162]: Shall we add the following comments like the other branches?
```
// Send the latest pod state we know to the watcher to make sure we didn't miss anything
```

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesClientApplication.scala [hunk 13]
```diff
@@ -133,29 +136,37 @@ private[spark] class Client(
...
+        watcher.eventReceived(Action.MODIFIED, podWithName.get())
+
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesClientApplication.scala:164]: Please add the following comment like the other branches.
```
// Break the while loop if the pod is completed or we don't want to wait
```

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesClientApplication.scala [hunk 14]
```diff
@@ -133,29 +136,37 @@ private[spark] class Client(
...
+        if(watcher.watchOrStop(sId)) {
+          logInfo(s"Stop watching as the pod has completed.")
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesClientApplication.scala:166]: Let's remove this. Otherwise this exists only at `branch-2.4`. People can be confused and think as a regression at branch-3.0 and `master`.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala [hunk 15]
```diff
@@ -28,8 +29,10 @@ import org.apache.spark.SparkException
 import org.apache.spark.internal.Logging
 import org.apache.spark.util.ThreadUtils
 
+
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala:32]: Let's remove this to minimize diff and to be consistent.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala [hunk 16]
```diff
@@ -77,9 +87,18 @@ private[k8s] class LoggingPodStatusWatcherImpl(
     }
   }
 
+  override def reset(): Unit = {
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala:90]: Shall we move this to line 69 (after `private def phase`) like the other branches?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala [hunk 17]
```diff
@@ -177,4 +190,34 @@ private[k8s] class LoggingPodStatusWatcherImpl(
...
+      }
+    }
+
+    if(podCompleted) {
+      logInfo(
+        pod.map { p => s"Container final statuses:\n\n${containersDescription(p)}" }
+          .getOrElse("No containers were found in the driver pod."))
+      logInfo(s"Application ${appId} with submission ID $sId finished")
+    } else {
+      logInfo(s"Got HTTP Gone code, resource version changed in k8s api. Creating a new watcher.")
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala:211]: It seems that we don't have this in the other branches. Please remove this.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala [hunk 18]
```diff
@@ -177,4 +190,34 @@ private[k8s] class LoggingPodStatusWatcherImpl(
...
+    }
+
+    logInfo(s"Watcher has stopped, pod completed status: ${podCompleted}")
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala:214]: Please remove this.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala [hunk 19]
```diff
@@ -177,4 +190,34 @@ private[k8s] class LoggingPodStatusWatcherImpl(
...
+
+    podCompleted
+  } else {
+    logInfo(s"Deployed Spark application ${appId} with submission ID $sId into Kubernetes")
+    logInfo(s"It seems we end up here, because we never want to wait for completion...")
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala:219]: Please remote this.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala [hunk 20]
```diff
@@ -177,4 +190,34 @@ private[k8s] class LoggingPodStatusWatcherImpl(
...
+      logInfo(s"Application ${appId} with submission ID $sId finished")
+    } else {
+      logInfo(s"Got HTTP Gone code, resource version changed in k8s api. Creating a new watcher.")
+    }
+
+    logInfo(s"Watcher has stopped, pod completed status: ${podCompleted}")
+
+    podCompleted
+  } else {
+    logInfo(s"Deployed Spark application ${appId} with submission ID $sId into Kubernetes")
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala:218]: Instead of `${appId}`, the other branches seem to use `${conf.appName}`, don't they?

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/submit/ClientSuite.scala [hunk 21]
```diff
@@ -205,6 +207,6 @@ class ClientSuite extends SparkFunSuite with BeforeAndAfter {
       loggingPodStatusWatcher,
       KUBERNETES_RESOURCE_PREFIX)
     submissionClient.run()
-    verify(loggingPodStatusWatcher).awaitCompletion()
+    verify(loggingPodStatusWatcher).watchOrStop("default:driver")
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/submit/ClientSuite.scala:210]: `default` -> `kubernetesConf.namespace()`?

**Review by dongjoon-hyun — COMMENTED** [review]
Thank you for updating, @jkleckner . I left some comments to match this to the other branches.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala [hunk 22]
```diff
@@ -177,4 +190,34 @@ private[k8s] class LoggingPodStatusWatcherImpl(
   private def formatTime(time: String): String = {
     if (time != null ||  time != "") time else "N/A"
   }
+
+  override def watchOrStop(sId: String): Boolean = if (waitForCompletion) {
+    logInfo(s"Waiting for application ${conf.appName} with submission ID $sId to finish...")
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala:195]: Oh, it's a compilation error.
```scala
[error] /home/jenkins/workspace/SparkPullRequestBuilder@2/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala:195: object appName is not a member of package conf
[error]     logInfo(s"Waiting for application ${conf.appName} with submission ID $sId to finish...")
[error]                                              ^
[error] one error found
```

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala [hunk 23]
```diff
@@ -77,9 +87,18 @@ private[k8s] class LoggingPodStatusWatcherImpl(
...
   }
 
+  override def reset(): Unit = {
+    resourceTooOldReceived = false
+  }
+
   override def onClose(e: KubernetesClientException): Unit = {
     logDebug(s"Stopping watching application $appId with last-observed phase $phase")
-    closeWatch()
+    if (e != null && e.getCode==HTTP_GONE) {
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala:96]: `e.getCode == HTTP_GONE`?

**Review by dongjoon-hyun — COMMENTED** [review]
For `LoggingPodStatusWatcherImpl`, I missed that it doesn't have `conf` which is different from the other branch `class LoggingPodStatusWatcherImpl(conf: KubernetesDriverConf)`. Sorry about that. Could you fix the compilation error?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala (hunk 7 above)
Comment (jkleckner) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala]: Done.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesClientApplication.scala (hunk 12 above)
Comment (jkleckner) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesClientApplication.scala:162]: Done

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesClientApplication.scala (hunk 13 above)
Comment (jkleckner) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesClientApplication.scala:164]: Done

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesClientApplication.scala (hunk 14 above)
Comment (jkleckner) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesClientApplication.scala:166]: Done

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala (hunk 15 above)
Comment (jkleckner) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala:32]: Done

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala (hunk 16 above)
Comment (jkleckner) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala:90]: Done

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala (hunk 17 above)
Comment (jkleckner) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala:211]: Done

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala (hunk 18 above)
Comment (jkleckner) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala:214]: Done

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala (hunk 19 above)
Comment (jkleckner) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala:219]: Done

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala (hunk 20 above)
Comment (jkleckner) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala:218]: Reverted to appId per below.

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/submit/ClientSuite.scala (hunk 21 above)
Comment (jkleckner) [reply resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/submit/ClientSuite.scala:210]: done

**Comment by jkleckner** [comment]
@dongjoon-hyun I made the changes.  No force-push this time.

Sorry to take so long but I did get very very busy.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala (hunk 23 above)
Comment (jkleckner) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala]: Missed this one somehow.  Updated now.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala (hunk 22 above)
Comment (jkleckner) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala]: Can you confirm that this is the way you want it?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesClientApplication.scala [hunk 24]
```diff
@@ -230,7 +242,9 @@ private[spark] class KubernetesClientApplication extends SparkApplication {
     val master = KubernetesUtils.parseMasterUrl(sparkConf.get("spark.master"))
     val loggingInterval = if (waitForAppCompletion) Some(sparkConf.get(REPORT_INTERVAL)) else None
 
-    val watcher = new LoggingPodStatusWatcherImpl(kubernetesAppId, loggingInterval)
+    val watcher = new LoggingPodStatusWatcherImpl(kubernetesAppId,
+      loggingInterval,
+      waitForAppCompletion)
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesClientApplication.scala:247]: Please revert this change. This is inconsistent with Apache Spark 3.1 and 3.0.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala [hunk 25]
```diff
@@ -42,13 +44,20 @@ private[k8s] trait LoggingPodStatusWatcher extends Watcher[Pod] {
  */
 private[k8s] class LoggingPodStatusWatcherImpl(
     appId: String,
-    maybeLoggingInterval: Option[Long])
+    maybeLoggingInterval: Option[Long],
+    waitForCompletion: Boolean)
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala:48]: Please revert this change. This is inconsistent with Apache Spark 3.1 and 3.0.

**Review by dongjoon-hyun — APPROVED** [review]
Thank you, @jkleckner and all.
+1, LGTM. Merged to branch-2.4 for Apache Spark 2.4.8.

**Comment by jkleckner** [comment]
Thank you, @dongjoon-hyun !
